### ISSUE TITLE
dataclass: Incompatible types in assignment to Optional[List[Optional[int]]]

### ISSUE THREAD

**Issue by Jasha10** [python/mypy#10636]
**Bug Report**

In dataclasses, using `default_factory` with an `Optional[List[Optional[int]]]` field produces an "Incompatible types in assignment" error

**To Reproduce**

```python
$ python --version
Python 3.9.5

$ mypy --version
mypy 0.902

$ cat tmp.py
from dataclasses import dataclass, field
from typing import Optional, List
@dataclass
class SubscriptedList:
    list: List[int] = field(default_factory=lambda: [1, 2])
    opt_list: Optional[List[int]] = field(default_factory=lambda: [1, 2])
    list_opt: List[Optional[int]] = field(default_factory=lambda: [1, 2])
    opt_list_opt: Optional[List[Optional[int]]] = field(default_factory=lambda: [1, 2])

$ mypy tmp.py
tmp.py:8: error: Incompatible types in assignment (expression has type "List[int]", variable has type "Optional[List[Optional[int]]]")
Found 1 error in 1 file (checked 1 source file)
```

This error does not come up for a normal variable assignment:
```python
var: Optional[List[Optional[int]]] = [1,2]  # mypy has no issue with this
```
If the default is `[1,2, None]` or `None`, no error is produced, but using `[None]` as the default does produce an error.

Note that a very similar error presents for `Optional[Dict[str, Optional[int]]]`:
```python
from dataclasses import dataclass, field
from typing import Dict, Optional
@dataclass
class SubscriptedDict:
    opt_dict_opt: Optional[Dict[str, Optional[int]]] = field(default_factory=lambda: {"foo": 123}) 
```
With the above, mypy gives:
```
tmp.py:5: error: Incompatible types in assignment (expression has type "Dict[str, int]", variable has type "Optional[Dict[str, Optional[int]]]")
```

**Expected Behavior**

I do not expect that assigning `default_factory=lambda: [1,2]` to an `Optional[List[Optional[int]]]` field produces an error.

**Your Environment**

- Mypy version used: 0.902
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.9.5
- Operating system and version: Ubuntu 18.04

**Comment by sobolevn** [comment]
Simplified reproduction:

```python
from typing import Optional, List

a = lambda: [1, 2]
reveal_type(a)

opt_list_opt: Optional[List[Optional[int]]] = a()
# error: Incompatible types in assignment (expression has type "List[int]", variable has type "Optional[List[Optional[int]]]")
```

**Comment by sobolevn** [comment]
@Jasha10 in the meantime, use `Sequence` if possible:

```python
from dataclasses import dataclass, field
from typing import Optional, List, Sequence

@dataclass
class SubscriptedList:
    list: Sequence[int] = field(default_factory=lambda: [1, 2])
    opt_list: Optional[Sequence[int]] = field(default_factory=lambda: [1, 2])
    list_opt: Sequence[Optional[int]] = field(default_factory=lambda: [1, 2])
    opt_list_opt: Optional[Sequence[Optional[int]]] = field(default_factory=lambda: [1, 2])

# ok
```
